src/network/lan.cpp

```diff
@@ -0,0 +1,335 @@
+/*
+Copyright (C) 2016 proller <proler@gmail.com>
+*/
+
+/*
+This file is part of Freeminer.
+
+Freeminer is free software: you can redistribute it and/or modify
+it under the terms of the GNU General Public License as published by
+the Free Software Foundation, either version 3 of the License, or
+(at your option) any later version.
+
+Freeminer  is distributed in the hope that it will be useful,
+but WITHOUT ANY WARRANTY; without even the implied warranty of
+MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+GNU General Public License for more details.
+
+You should have received a copy of the GNU General Public License
+along with Freeminer.  If not, see <http://www.gnu.org/licenses/>.
+*/
+
+#include "network/lan.h"
+#include <cstdint>
+#include "convert_json.h"
+#include "socket.h"
+#include "log.h"
+#include "settings.h"
+#include "version.h"
+#include "networkprotocol.h"
+#include "server/serverlist.h"
+#include "debug.h"
+#include "json/json.h"
+#include <mutex>
+#include <shared_mutex>
+#include "porting.h"
+#include "threading/thread.h"
+#include "network/address.h"
+
+//copypaste from ../socket.cpp
+#ifdef _WIN32
+#ifndef WIN32_LEAN_AND_MEAN
+#define WIN32_LEAN_AND_MEAN
+#endif
+// Without this some of the network functions are not found on mingw
+#ifndef _WIN32_WINNT
+#define _WIN32_WINNT 0x0501
+#endif
+#include <windows.h>
+#include <winsock2.h>
+#include <ws2tcpip.h>
+#define LAST_SOCKET_ERR() WSAGetLastError()
+typedef SOCKET socket_t;
+typedef int socklen_t;
+#else
+#include <sys/types.h>
+#include <sys/socket.h>
+#include <netinet/in.h>
+#include <fcntl.h>
+#include <netdb.h>
+#include <unistd.h>
+#include <arpa/inet.h>
+
+#ifndef __ANDROID__
+#include <ifaddrs.h>
+#define HAVE_IFADDRS 1
+#endif
+
+#define LAST_SOCKET_ERR() (errno)
+typedef int socket_t;
+#endif
+
+const char* adv_multicast_addr = "224.1.1.1";
+const static unsigned short int adv_port = 29998;
+static std::string ask_str;
+
+bool use_ipv6 = true;
+
+lan_adv::lan_adv() : Thread("lan_adv")
+{
+}
+
+void lan_adv::ask()
+{
+	if (!isRunning()) start();
+
+	if (ask_str.empty()) {
+		Json::Value j;
+		j["cmd"] = "ask";
+		ask_str = fastWriteJson(j);
+	}
+
+	send_string(ask_str);
+}
+
+void lan_adv::send_string(const std::string &str)
+{
+	if (g_settings->getBool("enable_ipv6")) {
+		std::vector<uint32_t> scopes;
+		// todo: windows and android
```

Comment:
I don't know. I didn't put that comment there. @proller did as far as I know. Ipv6 was in proller's code and I saw no reason to remove it, I just made it be dependent on the enable_ipv6 setting because that made sense to me.